### PR TITLE
Updated Kafka message

### PR DESCRIPTION
**Pull request by jeongkyun-oh**
## Proposed changes

- In order not to mix the messages from multi-producers, the message format is updated.
- Introduced version/producerId for the message
  - the version is added for backward-compatibility so that the old message production and consumption should work as normal.
  - the producerId is added to distinguish the messages with the same key, but from different data source.

<p align="center"></p>

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: datasync/chaindatafetcher/kafka/config.go [hunk 1]
```diff
@@ -42,10 +44,14 @@ const (
...
 	DefaultSegmentSizeBytes     = 1000000 // 1 MB
 	DefaultMaxMessageNumber     = 100     // max number of messages in buffer
+	DefaultKafkaMessageVersion  = MsgVersion1_0
+	DefaultProducerIdPrefix     = "producer-"
 )
 
 type KafkaConfig struct {
 	SaramaConfig         *sarama.Config `json:"-"` // kafka client configurations.
+	MsgVersion           string         // MsgVersion is the version of Kafka message.
+	ProducerId           string         // ProducerId is for the identification of the message publisher.
```
Comment (aidan-kwon) [inline datasync/chaindatafetcher/kafka/config.go:54]: `String` method doesn't print new fields.

Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/config.go:54]: The method is updated. Thanks.

File: datasync/chaindatafetcher/kafka/kafka.go [hunk 2]
```diff
@@ -30,12 +30,21 @@ const (
 	// item indices of message header
 	MsgHeaderTotalSegments = iota
 	MsgHeaderSegmentIdx
+	MsgHeaderVersion
+	MsgHeaderProducerId
 	MsgHeaderLength
+	LegacyMsgHeaderLength = 2
```
Comment (aidan-kwon) [inline datasync/chaindatafetcher/kafka/kafka.go:36]: Why don't you separate this value from iota constians?

File: datasync/chaindatafetcher/kafka/kafka.go [hunk 3]
```diff
@@ -164,7 +188,7 @@ func (k *Kafka) Publish(topic string, data interface{}) error {
 	}
 	segments, totalSegments := k.split(dataBytes)
 	for idx, segment := range segments {
-		msg := k.makeProducerMessage(topic, key, segment, uint64(idx), uint64(totalSegments))
+		msg := k.makeProducerMessage(topic, key, k.config.MsgVersion, k.config.ProducerId, segment, uint64(idx), uint64(totalSegments))
```
Comment (aidan-kwon) [inline datasync/chaindatafetcher/kafka/kafka.go:191]: `makeProducerMessage` doesn't need to pass `.kconfig.MsgVersion` and `k.config.ProducerId` because it  can access to the config also.

File: datasync/chaindatafetcher/kafka/kafka_test.go [hunk 4]
```diff
@@ -141,14 +166,19 @@ func (s *KafkaSuite) TestKafka_CreateAndDeleteTopic() {
 }
 
 type kafkaData struct {
-	Data []byte `json:"data"`
+	Number int
```
Comment (aidan-kwon) [inline datasync/chaindatafetcher/kafka/kafka_test.go:169]: Why did you add `Number`?

Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/kafka_test.go:169]: This is for testing purpose. I wanted to check whether or not the segments are combined well with the given key and given producerId. I put `Number` because it has `Key()` method.

Comment (aidan-kwon) [reply datasync/chaindatafetcher/kafka/kafka_test.go:169]: Okay. Thanks

File: datasync/chaindatafetcher/kafka/kafka.go (hunk 2 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/kafka.go]: I made separate constant for `LegacyMsgHeaderLength`.

File: datasync/chaindatafetcher/kafka/kafka.go (hunk 3 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/kafka.go]: You are right. I have updated. Thanks.

**Review by aidan-kwon — DISMISSED** [review]
LGTM

File: datasync/chaindatafetcher/kafka/config.go [hunk 5]
```diff
@@ -74,14 +80,22 @@ func GetDefaultKafkaConfig() *KafkaConfig {
...
 		MaxMessageNumber:     DefaultMaxMessageNumber,
+		MsgVersion:           DefaultKafkaMessageVersion,
+		ProducerId:           GetDefaultProducerId(),
 	}
 }
 
+func GetDefaultProducerId() string {
+	rb := common.MakeRandomBytes(8)
+	randomString := hexutil.Encode(rb)
+	return DefaultProducerIdPrefix + randomString[2:]
```
Comment (yoomee1313) [inline datasync/chaindatafetcher/kafka/config.go:91]: Can you explain why you use randomString[2:]? I'm just curious.

**Review by yoomee1313 — DISMISSED** [review]
It looks good to me

File: datasync/chaindatafetcher/kafka/consumer.go [hunk 6]
```diff
@@ -65,8 +68,30 @@ func newSegment(msg *sarama.ConsumerMessage) (*Segment, error) {
...
-		return nil, fmt.Errorf("%v [header length: %v]", wrongHeaderNumberErrorMsg, len(msg.Headers))
+	headerLen := len(msg.Headers)
+	if headerLen != MsgHeaderLength && headerLen != LegacyMsgHeaderLength {
+		return nil, fmt.Errorf("%v [header length: %v]", wrongHeaderNumberErrorMsg, headerLen)
+	}
+
+	version := ""
+	producerId := ""
+
+	if len(msg.Headers) == MsgHeaderLength {
```
Comment (hyochan-brown) [inline datasync/chaindatafetcher/kafka/consumer.go:79]: How do you think about using headerLen?

File: datasync/chaindatafetcher/kafka/config.go (hunk 5 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/config.go:91]: @yoomee1313 In order to give an unique producer id, I put random hex string without prefix `0x`.

Comment (yoomee1313) [reply datasync/chaindatafetcher/kafka/config.go:91]: understood, thanks!

File: datasync/chaindatafetcher/kafka/consumer.go [hunk 7]
```diff
@@ -65,8 +68,30 @@ func newSegment(msg *sarama.ConsumerMessage) (*Segment, error) {
...
+		keyVersion := string(msg.Headers[MsgHeaderVersion].Key)
```
Comment (hyochan-brown) [inline datasync/chaindatafetcher/kafka/consumer.go:80]: I may have misunderstood.
I think it have a dependency on the header record order.
Is it okay..?

**Review by hyochan-brown — COMMENTED** [review]
LGTM
I have a few questions!

File: datasync/chaindatafetcher/kafka/consumer.go (hunk 6 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/consumer.go]: Reflected, thanks.

File: datasync/chaindatafetcher/kafka/consumer.go (hunk 7 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/consumer.go:80]: I agree with you. The order or the header is also checked previously as well, but it is too strict to check. 
However, let me update about this later.

**Review by hyochan-brown — APPROVED** [review]

File: datasync/chaindatafetcher/kafka/consumer.go (hunk 7 above)
Comment (hyochan-brown) [reply datasync/chaindatafetcher/kafka/consumer.go:80]: Thank you for your kind answer!

**Review by yoomee1313 — APPROVED** [review]

**Review by aidan-kwon — APPROVED** [review]

**Review by KimKyungup — APPROVED** [review]
